server/rating_service/rating_service.py

```diff
@@ -0,0 +1,254 @@
+from typing import Dict
+from .typedefs import RatingData, GameRatingData, GameRatingSummary, PlayerID
+
+import asyncio
+
+from server.db import FAFDatabase
+from server.player_service import PlayerService
+from server.decorators import with_logger
+from server.metrics import rating_service_backlog
+
+from server.games.game_results import GameOutcome
+
+from server.rating import RatingType
+from trueskill import Rating
+from .game_rater import GameRater, GameRatingError
+
+from sqlalchemy import select
+from server.db.models import ladder1v1_rating as ladder1v1_table
+from server.db.models import global_rating as global_table
+
+
+class RatingServiceError(Exception):
+    pass
+
+
+class ServiceNotReadyError(RatingServiceError):
+    pass
+
+
+class RatingNotFoundError(RatingServiceError):
+    pass
+
+
+@with_logger
+class RatingService:
+    """
+    Service responsible for calculating and saving trueskill rating updates.
+    To avoid race conditions, rating updates from a single game ought to be
+    atomic.
+    """
+
+    def __init__(self, database: FAFDatabase, player_service: PlayerService):
+        self._db = database
+        self._player_service_callback = player_service.signal_player_rating_change
+        self._accept_input = False
+        self._queue = asyncio.Queue()
+        self._task = None
+
+    async def initialize(self):
+        # TODO use create_task in python3.7
+        self._accept_input = True
+        self._logger.debug("RatingService starting...")
+        self._task = asyncio.ensure_future(self._handle_rating_queue())
+
+    async def enqueue(self, summary: GameRatingSummary) -> None:
+        if not self._accept_input:
+            self._logger.warning("Dropped rating request %s", summary)
+            raise ServiceNotReadyError(
+                "RatingService not yet initialized or shutting down."
+            )
+
+        self._logger.debug("Queued up rating request %s", summary)
+        await self._queue.put(summary)
+        rating_service_backlog.set(self._queue.qsize())
+
+    async def _handle_rating_queue(self):
+        self._logger.info("RatingService started!")
+        while True:
+            summary = await self._queue.get()
+            self._logger.debug("Now rating request %s", summary)
+
+            try:
+                await self._rate(summary)
+            except GameRatingError:
+                self._logger.warning("Error rating game %s", summary)
+            except RatingNotFoundError:
+                self._logger.warning("Missing rating entry to rate game %s", summary)
+            else:
+                self._logger.debug("Done rating request.")
+
+            self._queue.task_done()
+            rating_service_backlog.set(self._queue.qsize())
+
+            if self._queue.empty() and not self._accept_input:
+                self._logger.info("RatingService shutting down.")
+                break
```

Comment:
This isn't actually necessary, since `shutdown` will kill the task once the queue is empty. Might be cleaner if these lines here are removed.

Reply:
Ah, of course, thanks!